### DISCUSSION TITLE
MSA application default options

### DISCUSSION THREAD

**Discussion by EiriniKot** [biotite-dev/biotite#501] [Q&A] [answered]
Do you know which are the default options for biotite - mafft, clustal-omega and muscle? ( Or where can I find them)

**Comment by padix-key** [comment, marked as answer]
Hi. Like all `LocalApp` subclasses, The MSA software interfaces have the [`get_command()`](https://www.biotite-python.org/apidoc/biotite.application.mafft.MafftApp.html?highlight=mafftapp#biotite.application.mafft.MafftApp.get_command) method, that returns the executed command, including all used options.

**Comment by EiriniKot** [comment]
I am using this object :biotite.application.mafft.MafftApp (https://www.biotite-python.org/apidoc/biotite.application.mafft.MafftApp.html)

```
aligner_app
PyDev console: starting.
<class 'biotite.application.mafft.MafftApp'>


aligner_app.get_command()
Traceback (most recent call last):
  File "/snap/pycharm-professional/359/plugins/python/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
  File "/home/eirini/anaconda3/envs/gaz-torch-cuda/lib/python3.8/site-packages/biotite/application/application.py", line 52, in wrapper
    instance = args[0]
IndexError: tuple index out of range
```

**Reply by padix-key** [reply]
From your output it seems you called `get_command()` on the `MafftApp` class instead of an `MafftApp` object. However, `get_method()` is an instance method, and it must be, since the command depends on how the created `MafftApp` object is configured. So this should work:

```python
seq1 = ProteinSequence("BIQTITE")
seq2 = ProteinSequence("TITANITE")
seq3 = ProteinSequence("BISMITE")
seq4 = ProteinSequence("IQLITE")
app = MafftApp([seq1, seq2, seq3, seq4])
app.start()
app.join()
print(app.get_command())
```

Nonetheless, I admit a clearer error message would be helpful here.

**Comment by EiriniKot** [comment]
I am using align function to do the alignment instead of `app = MafftApp([seq1, seq2, seq3, seq4])` so basically I am creating the app instance from MafftApp and then I do `app.align([seq1, seq2, seq3, seq4])`

**Reply by padix-key** [reply]
Could you paste your code? The output shown above indicates that `aligner_app` is the class `MafftApp` and not an object:

> <class 'biotite.application.mafft.MafftApp'>

Hence, calling `aligner_app.get_command()` fails.

**Reply by EiriniKot** [reply]
```
aligner_app = mafft.MafftApp
aligner_app.align([seq1,seq2,seq3,seq4])
Alignment([ProteinSequence("BIQTITE"), ProteinSequence("TITANITE"), ProteinSequence("BISMITE"), ProteinSequence("IQLITE")], np.array([[-1,  0, -1, -1],
       [ 0,  1,  0, -1],
       [ 1,  2,  1,  0],
       [ 2,  3,  2,  1],
       [ 3,  4,  3,  2],
       [ 4,  5,  4,  3],
       [ 5,  6,  5,  4],
       [ 6,  7,  6,  5]]), score=None)
aligner_app.get_command()
Traceback (most recent call last):
  File "/snap/pycharm-professional/359/plugins/python/helpers/pydev/pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
  File "/home/eirini/anaconda3/envs/gaz-torch-cuda/lib/python3.8/site-packages/biotite/application/application.py", line 52, in wrapper
    instance = args[0]
IndexError: tuple index out of range
```

**Reply by padix-key** [reply]
The problem is the first line:

> aligner_app = mafft.MafftApp

`aligner_app` is assigned to the class, not to an object of it. This is OK for simply aligning, since you use the shortcut static method `align()` that is called on the class itself. However, to be able to get the command you need to run it based on a created object. So this would be the modification:

```python
aligner_app = mafft.MafftApp([seq1,seq2,seq3,seq4])
aligner_app.start()
aligner_app.join()
print(aligner_app.get_command())
print(aligner_app.get_alignment())
```

I would recommend the object-based API, if specific parameters should be given to an `Application` or side results (such as the command) should be obtained from the application run.

**Reply by EiriniKot** [reply]
thank you
